client/client.go

```diff
@@ -177,36 +282,109 @@ func (pb *ProofBuilder) fetchNodes(ctx context.Context, nodes proof.Nodes) ([][]
 	return hashes, nil
 }
 
-// FetchRangeNodes returns the set of nodes representing the compact range covering
-// a log of size s.
-func FetchRangeNodes(ctx context.Context, s uint64, f TileFetcherFunc) ([][]byte, error) {
-	nc := newNodeCache(f, s)
-	nIDs := make([]compact.NodeID, 0, compact.RangeSize(0, s))
-	nIDs = compact.RangeNodes(0, s, nIDs)
-	hashes := make([][]byte, 0, len(nIDs))
-	for _, n := range nIDs {
-		h, err := nc.GetNode(ctx, n)
+// LogStateTracker represents a client-side view of a target log's state.
+// This tracker handles verification that updates to the tracked log state are
+// consistent with previously seen states.
+type LogStateTracker struct {
+	origin              string
+	consensusCheckpoint ConsensusCheckpointFunc
+	cpSigVerifier       note.Verifier
+	tileFetcher         TileFetcherFunc
+
+	// The fields under here will all be updated at the same time.
+	// Access to any of these fields is guarded by mu.
+	mu sync.RWMutex
+
+	// latestConsistent is the deserialised form of LatestConsistentRaw
+	latestConsistent log.Checkpoint
+	// latestConsistentRaw holds the raw bytes of the latest proven-consistent
+	// LogState seen by this tracker.
+	latestConsistentRaw []byte
+	// proofBuilder for building proofs at LatestConsistent checkpoint.
+	proofBuilder *ProofBuilder
+}
+
+// NewLogStateTracker creates a newly initialised tracker.
+// If a serialised LogState representation is provided then this is used as the
+// initial tracked state, otherwise a log state is fetched from the target log.
+func NewLogStateTracker(ctx context.Context, cpF CheckpointFetcherFunc, tF TileFetcherFunc, checkpointRaw []byte, nV note.Verifier, origin string, cc ConsensusCheckpointFunc) (*LogStateTracker, error) {
+	ret := &LogStateTracker{
+		origin:              origin,
+		consensusCheckpoint: cc,
+		cpSigVerifier:       nV,
+		tileFetcher:         tF,
+	}
+	if len(checkpointRaw) > 0 {
+		ret.latestConsistentRaw = checkpointRaw
+		cp, _, _, err := log.ParseCheckpoint(checkpointRaw, origin, nV)
 		if err != nil {
-			return nil, err
+			return ret, err
```

Comment:
I'm wondering if this is safe/useful to be _able_ to continue using it if the provided checkpoint is borked? Particularly given we're setting `latestConsistentRaw` to the borked checkpoint.